### PR TITLE
Error functions

### PR DESCRIPTION
**Pull request by Firgeis**
This PR should solve  #108

* Bindings for ERR_get_error and ERR_peek_last_error that return the raw errors as a tuple
* Replaces the function used for exceptions with last_error so it doesn't pop the error queue inadvertently

For compatibility I have reproduced the behavior of the function used to print exceptions with the ERR_error_string_n but with the functionality of the ERR_peek_last_error  function (peeks the last error)

### CODE REVIEW

File: src/ssl.mli [hunk 1]
```diff
@@ -204,6 +204,14 @@ val init : ?thread_safe:bool -> unit -> unit
     occurred. *)
 val get_error_string : unit -> string
 
+(** Retrieve the earliest error from the error queue then it removes the entry.
+    Returns the code and library and reason strings *)
+val get_error : unit -> int * string * string
+
+(** Retrieves the latest error code from the thread's error queue without
+    modifying it. Returns the code and library and reason strings. *)
+val peek_error_last : unit -> int * string * string
```
Comment (anmonteiro) [inline src/ssl.mli:213]: can we call it `peek_last_error` like the SSL call (https://www.openssl.org/docs/man3.1/man3/ERR_get_error.html)?

should we also provide a binding to `peek_error`?

File: src/ssl.mli [hunk 2]
```diff
@@ -204,6 +204,14 @@ val init : ?thread_safe:bool -> unit -> unit
     occurred. *)
 val get_error_string : unit -> string
```
Comment (anmonteiro) [inline src/ssl.mli:205]: should we mention that this function advances the error queue?

File: src/ssl_stubs.c [hunk 3]
```diff
@@ -262,7 +263,55 @@ CAMLprim value ocaml_ssl_get_error_string(value unit)
   CAMLreturn(caml_copy_string(buf));
 }
 
+CAMLprim value ocaml_ssl_get_error_struct(value unit)
+{
+  CAMLparam1(unit);
+  unsigned long code = ERR_get_error();
+  CAMLlocal3(result, libstring, reasonstring);
```
Comment (anmonteiro) [inline src/ssl_stubs.c:270]: nit: swap to keep the 2 CAML-related lines together?

```suggestion
  CAMLlocal3(result, libstring, reasonstring);
  unsigned long code = ERR_get_error();
```

File: src/ssl_stubs.c [hunk 4]
```diff
@@ -262,7 +263,55 @@ CAMLprim value ocaml_ssl_get_error_string(value unit)
...
+  Store_field(result, 2, reasonstring);
+
+  CAMLreturn(result);
+}
 
+CAMLprim value ocaml_ssl_peek_error_last_struct(value unit)
+{
+  CAMLparam1(unit);
+  unsigned long code = ERR_peek_last_error();
+  CAMLlocal3(result, libstring, reasonstring);
```
Comment (anmonteiro) [inline src/ssl_stubs.c:295]:
```suggestion
  CAMLlocal3(result, libstring, reasonstring);
  unsigned long code = ERR_peek_last_error();
```

File: src/ssl.mli (hunk 1 above)
Comment (Firgeis) [reply src/ssl.mli]: Not sure, you think it would be useful?

File: src/ssl.ml [hunk 5]
```diff
@@ -81,6 +81,22 @@ type bigarray = (char, Bigarray.int8_unsigned_elt, Bigarray.c_layout) Bigarray.A
 
 external get_error_string : unit -> string = "ocaml_ssl_get_error_string"
 
+external error_struct : int -> int * string * string = "ocaml_ssl_error_struct"
+
+let get_error () =
+  error_struct 0
```
Comment (anmonteiro) [inline src/ssl.ml:87]: this is nice. now you can make it a variant:

```ocaml
type err_function = | Get_error | Peek_error | Peek_last_error
```

File: src/ssl.ml [hunk 6]
```diff
@@ -81,6 +81,24 @@ type bigarray = (char, Bigarray.int8_unsigned_elt, Bigarray.c_layout) Bigarray.A
 
 external get_error_string : unit -> string = "ocaml_ssl_get_error_string"
 
+type err_function = | Get_error | Peek_error | Peek_last_error
+
+external error_struct : err_function -> int * string * string = "ocaml_ssl_error_struct"
```
Comment (anmonteiro) [inline src/ssl.ml:86]: How about we provide a `Ssl.Error` module with `type t = { code: int; lib: string; reason: string }`?

Comment (anmonteiro) [reply src/ssl.ml:86]: can even expose it in the `.mli` as:

```ocaml
module Error : sig
  type t = private { code: int; lib: string; reason: string } 
end
```

to make sure the library is the only one constructing it.

File: src/ssl_stubs.c [hunk 7]
```diff
@@ -262,7 +263,41 @@ CAMLprim value ocaml_ssl_get_error_string(value unit)
...
+      code = ERR_peek_last_error();
+      break;
+  }
+  result = caml_alloc_tuple(3);
+  const char *lib = ERR_lib_error_string(code);
+  const char *reason = ERR_reason_error_string(code);
+  if (lib != NULL) {
+    libstring = caml_copy_string(lib);
+  } else {
+    libstring = caml_copy_string("lib(0)");
```
Comment (anmonteiro) [inline src/ssl_stubs.c:288]: how about making `lib` and `reason` `string option`s instead of making up these strings?

File: src/ssl.mli [hunk 8]
```diff
@@ -200,10 +200,26 @@ exception Verify_error of verify_error
     [Ssl_threads.init] first. *)
 val init : ?thread_safe:bool -> unit -> unit
 
-(** Retrieve a human-readable message that corresponds to the last error that
-    occurred. *)
+(** Retrieve a human-readable message that corresponds to the earliest error
+    code from the thread's error queue and removes the entry. *)
 val get_error_string : unit -> string
```
Comment (anmonteiro) [inline src/ssl.mli:205]:
```suggestion
val get_error_string : unit -> string
  [@@ocaml.alert deprecated "Use [Ssl.Error.get_error] instead"]
```

File: src/ssl.mli [hunk 9]
```diff
@@ -200,9 +200,26 @@ exception Verify_error of verify_error
...
+  [@@ocaml.alert deprecated "Use [Ssl.Error.get_error] instead"]
+
+module Error: sig
```
Comment (anmonteiro) [inline src/ssl.mli:208]: @Firgeis In a follow-up PR, it'd be nice to explain in this doc that OpenSSL keeps an error queue per thread, that it gets cleaned up before every TLS operation, which can then lead to users of ocaml-ssl understanding what it means to get vs peek errors.

**Comment by anmonteiro** [comment]
Thank you

File: src/ssl.mli (hunk 9 above)
Comment (Firgeis) [reply src/ssl.mli:208]: @anmonteiro  When you say in this doc, you mean as a comment for the error module?

Comment (anmonteiro) [reply src/ssl.mli:208]: yeah, in this `.mli` file.

**Pull request merged**